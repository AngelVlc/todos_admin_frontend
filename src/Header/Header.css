.Header {
    background-color: #282c34;
    display: flex;
    flex-direction: column;
    align-items: left;
    justify-content: center;
    font-size: calc(8px + 1vmin);
    color: white;
    padding-left: 50px;
}