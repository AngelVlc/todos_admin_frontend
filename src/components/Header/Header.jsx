import React, { useRef } from 'react';
import { useHistory } from 'react-router-dom';

export const Header = () => {
    const navBarMenuRef = useRef();
    const navBarBurguerRef = useRef();
    let history = useHistory();
    const userInfoStr = localStorage.getItem('userInfo');
    let userInfo = userInfoStr ? JSON.parse(userInfoStr) : null;

    const onLogoutClick = () => {
        localStorage.setItem('userInfo', null);
        history.push('/login');
    }

    const onToggleNavBar = (e) => {
        navBarMenuRef.current.classList.toggle('is-active');
        navBarBurguerRef.current.classList.toggle('is-active');
    }

    return (
        <nav className="navbar is-dark">
            <div className="navbar-brand">
                <span className="navbar-item is-size-3">To Dos</span>
                {userInfo &&
                    <div className="navbar-burger" ref={navBarBurguerRef} onClick={(e) => onToggleNavBar(e)}>
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                    </div>
                }
            </div>
            <div className="navbar-menu" ref={navBarMenuRef}>
                <div className="navbar-start">
                    {userInfo &&
                        <>
                            <span className="navbar-item" onClick={() => history.push('/')}>Home</span>
                            <span className="navbar-item" onClick={() => history.push('/lists')}>Lists</span>
                        </>
                    }
                    {userInfo?.isAdmin &&
                        <span className="navbar-item" onClick={() => history.push('/users')}>Users</span>
                    }
                </div>
            </div>
            {userInfo &&
                <div className="navbar-end">
                    <div className="navbar-item">
                        <div className="buttons">
                            <button className="button is-light is-small" data-testid="logOut" onClick={() => onLogoutClick()}>
                                <span>{userInfo.userName}</span>
                                <span className="icon is-small">
                                    <i className="fas fa-sign-out-alt"></i>
                                </span>
                            </button>
                        </div>
                    </div>
                </div>
            }
        </nav>
    )
}
