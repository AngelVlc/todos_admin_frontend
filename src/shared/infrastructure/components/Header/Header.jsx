import React, { useRef, useContext } from 'react';
import { AppContext } from '../../contexts';
import { useHistory } from 'react-router-dom';

export const Header = () => {
    const { auth } = useContext(AppContext)
    const navBarMenuRef = useRef();
    const navBarBurguerRef = useRef();
    let history = useHistory();

    const onLogoutClick = () => {
        history.push('/login');
    }

    const onToggleNavBar = (e) => {
        navBarMenuRef.current.classList.toggle('is-active');
        navBarBurguerRef.current.classList.toggle('is-active');
    }

    return (
        <nav className="navbar is-dark">
            <div className="navbar-brand">
                <span className="navbar-item is-size-3">To Dos</span>
                {auth.info &&
                    <div className="navbar-burger" ref={navBarBurguerRef} onClick={(e) => onToggleNavBar(e)}>
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                    </div>
                }
            </div>
            <div className="navbar-menu" ref={navBarMenuRef}>
                <div className="navbar-start">
                    {auth.info &&
                        <>
                            <span className="navbar-item" data-testid="goToRoot" onClick={() => history.push('/')}>Home</span>
                            <span className="navbar-item" data-testid="goToLists" onClick={() => history.push('/lists')}>Lists</span>
                        </>
                    }
                    {auth.info?.isAdmin &&
                        <span className="navbar-item" data-testid="goToUsers" onClick={() => history.push('/users')}>Users</span>
                    }
                </div>
            </div>
            {auth.info &&
                <div className="navbar-end">
                    <div className="navbar-item">
                        <div className="buttons">
                            <button className="button is-light is-small" data-testid="logOut" onClick={() => onLogoutClick()}>
                                <span>{auth.info.userName}</span>
                                <span className="icon is-small">
                                    <i className="fas fa-sign-out-alt"></i>
                                </span>
                            </button>
                        </div>
                    </div>
                </div>
            }
        </nav>
    )
}
