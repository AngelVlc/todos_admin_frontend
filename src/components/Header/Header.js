import React, { useContext } from 'react';
import { AuthContext } from '../../contexts/AuthContext';
import { doLogout } from '../../actions/auth';
import './Header.css';

export const Header = () => {
    const { auth, dispatch } = useContext(AuthContext)

    const logoutClick = (dispatch) => {
        dispatch(doLogout());
    }

    return (
        <header className="Header">
            <h1>To Dos Admin</h1>
            {auth.info &&
                <>
                    <h4>User: {auth.info.userName}</h4>
                    <button onClick={() => logoutClick(dispatch)}>Log out</button>
                </>
            }
        </header>
    )
}
