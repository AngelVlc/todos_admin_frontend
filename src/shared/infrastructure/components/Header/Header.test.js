import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { Router } from "react-router-dom";
import { Header } from "./Header";
import { AppContext } from "../../contexts";
import { createMemoryHistory } from "history";

const mockAuthDispatch = jest.fn();
const mockHistoryPush = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const renderWithRouterAndContext = (auth) => {
  const history = createMemoryHistory();
  const context = { authDispatch: mockAuthDispatch, auth };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <Router history={history}>{<Header />}</Router>
      </AppContext.Provider>
    ),
  };
};

afterEach(cleanup);

it("should match the snapshot when the user is not logged in", () => {
  const { asFragment } = renderWithRouterAndContext({ info: null });

  expect(asFragment()).toMatchSnapshot();
});

it("should match the snapshot when the user is logged in", () => {
  const { asFragment } = renderWithRouterAndContext({
    info: { userName: "user" },
  });

  expect(asFragment()).toMatchSnapshot();
});

it("should match the snapshot when an admin user is logged in", () => {
  const { asFragment } = renderWithRouterAndContext({
    info: { userName: "admin", isAdmin: true },
  });

  expect(asFragment()).toMatchSnapshot();
});

it("should do logout", async () => {
  const { getByTestId } = renderWithRouterAndContext({
    info: { userName: "user" },
  });

  await waitFor(() => {
    fireEvent.click(getByTestId("logOut"));
  });

  expect(mockHistoryPush).toHaveBeenCalled();
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/login");
});

it("should go to root", async () => {
  const { getByTestId } = renderWithRouterAndContext({
    info: { userName: "user" },
  });

  await waitFor(() => {
    fireEvent.click(getByTestId("goToRoot"));
  });

  expect(mockHistoryPush).toHaveBeenCalled();
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/");
});

it("should go to lists", async () => {
  const { getByTestId } = renderWithRouterAndContext({
    info: { userName: "user" },
  });

  await waitFor(() => {
    fireEvent.click(getByTestId("goToLists"));
  });

  expect(mockHistoryPush).toHaveBeenCalled();
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists");
});

it("should go to users", async () => {
  const { getByTestId } = renderWithRouterAndContext({
    info: { userName: "user", isAdmin: true },
  });

  await waitFor(() => {
    fireEvent.click(getByTestId("goToUsers"));
  });

  expect(mockHistoryPush).toHaveBeenCalled();
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/users");
});
