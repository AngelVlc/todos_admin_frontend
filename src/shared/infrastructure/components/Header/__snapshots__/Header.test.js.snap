// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot when an admin user is logged in 1`] = `
<DocumentFragment>
  <nav
    class="navbar is-dark"
  >
    <div
      class="navbar-brand"
    >
      <span
        class="navbar-item is-size-3"
      >
        To Dos
      </span>
      <div
        class="navbar-burger"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </div>
    </div>
    <div
      class="navbar-menu"
    >
      <div
        class="navbar-start"
      >
        <span
          class="navbar-item is-clickable"
          data-testid="goToRoot"
        >
          Home
        </span>
        <span
          class="navbar-item is-clickable"
          data-testid="goToLists"
        >
          Lists
        </span>
        <span
          class="navbar-item is-clickable"
          data-testid="goToUsers"
        >
          Users
        </span>
      </div>
      <div
        class="navbar-end"
      >
        <hr
          class="my-0"
        />
        <span
          class="navbar-item is-clickable"
          data-testid="logOut"
        >
          Log Out admin
        </span>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Header should match the snapshot when the user is logged in 1`] = `
<DocumentFragment>
  <nav
    class="navbar is-dark"
  >
    <div
      class="navbar-brand"
    >
      <span
        class="navbar-item is-size-3"
      >
        To Dos
      </span>
      <div
        class="navbar-burger"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </div>
    </div>
    <div
      class="navbar-menu"
    >
      <div
        class="navbar-start"
      >
        <span
          class="navbar-item is-clickable"
          data-testid="goToRoot"
        >
          Home
        </span>
        <span
          class="navbar-item is-clickable"
          data-testid="goToLists"
        >
          Lists
        </span>
      </div>
      <div
        class="navbar-end"
      >
        <hr
          class="my-0"
        />
        <span
          class="navbar-item is-clickable"
          data-testid="logOut"
        >
          Log Out user
        </span>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Header should match the snapshot when the user is not logged in 1`] = `
<DocumentFragment>
  <nav
    class="navbar is-dark"
  >
    <div
      class="navbar-brand"
    >
      <span
        class="navbar-item is-size-3"
      >
        To Dos
      </span>
    </div>
    <div
      class="navbar-menu"
    >
      <div
        class="navbar-start"
      />
    </div>
  </nav>
</DocumentFragment>
`;
